src/runtime-rs/crates/resource/src/cgroups/mod.rs

```diff
@@ -0,0 +1,213 @@
+// Copyright (c) 2019-2022 Alibaba Cloud
+// Copyright (c) 2019-2022 Ant Group
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+mod utils;
+
+use std::{
+    collections::{HashMap, HashSet},
+    iter::FromIterator,
+    sync::Arc,
+};
+
+use anyhow::{anyhow, Result};
+use cgroups_rs::{cgroup_builder::CgroupBuilder, Cgroup, CgroupPid, CpuResources, Resources};
+use hypervisor::Hypervisor;
+use kata_sys_util::spec::load_oci_spec;
+use kata_types::config::TomlConfig;
+use oci::LinuxResources;
+use tokio::sync::RwLock;
+
+pub struct CgroupConfig {
+    pub path: String,
+    pub overhead_path: String,
+    pub sandbox_cgroup_only: bool,
+}
+
+impl CgroupConfig {
+    fn new(sid: &str, toml_config: &TomlConfig) -> Result<Self> {
+        let overhead_path = utils::gen_overhead_path(sid);
+        let spec = load_oci_spec()?;
+        let path = spec
+            .linux
+            // The trim of '/' is important, because cgroup_path is a relative path.
+            .map(|linux| linux.cgroups_path.trim_start_matches('/').to_string())
+            .unwrap_or_default();
+
+        Ok(Self {
+            path,
+            overhead_path,
+            sandbox_cgroup_only: toml_config.runtime.sandbox_cgroup_only,
+        })
+    }
+}
+
+pub struct CgroupsResource {
+    resources: Arc<RwLock<HashMap<String, Resources>>>,
+    cgroup_manager: Cgroup,
+    overhead_cgroup_manager: Option<Cgroup>,
+}
+
+impl CgroupsResource {
+    pub fn new(sid: &str, toml_config: &TomlConfig) -> Result<Self> {
+        let config = CgroupConfig::new(sid, toml_config)?;
+
+        // Create the sandbox cgroups manager (cgroups on Linux).
+        // Depending on the sandbox_cgroup_only value, this cgroup
+        // will either hold all the pod threads (sandbox_cgroup_only is true)
+        // or only the virtual CPU ones (sandbox_cgroup_only is false).
+        let hier = cgroups_rs::hierarchies::auto();
+        let cgroup_manager = CgroupBuilder::new(&config.path).build(hier);
+
+        // The shim configuration is requesting that we do not put all threads
+        // into the sandbox resource controller.
+        // We're creating an overhead controller, with no constraints. Everything but
+        // the vCPU threads will eventually make it there.
+        let overhead_cgroup_manager = if !config.sandbox_cgroup_only {
+            let hier = cgroups_rs::hierarchies::auto();
+            Some(CgroupBuilder::new(&config.overhead_path).build(hier))
+        } else {
+            None
+        };
+
+        // Add the runtime to the VMM sandbox resource controller
+
+        // By adding the runtime process to either the sandbox or overhead controller, we are making
+        // sure that any child process of the runtime (i.e. *all* processes serving a Kata pod)
+        // will initially live in this controller. Depending on the sandbox_cgroup settings, we will
+        // then move the vCPU threads between resource controllers.
+        overhead_cgroup_manager
```

Comment:
Here is something tricky. suggest to:

```rust
if config.sandbox_cgroup_only {
    cgroup_manager.add_task_by_tgid()
} else {
    overhead_cgroup_manager.add_task_by_tgid()
}
```

Reply:
Thanks for your review, already updated.